File: product-service/src/main/java/com/intexsoft/webshop/productservice/mapper/AttributeMapper.java
package com.intexsoft.webshop.productservice.mapper;

import com.intexsoft.webshop.productservice.dto.attribute.AttributeCreateDto;
import com.intexsoft.webshop.productservice.dto.attribute.AttributeDto;
import com.intexsoft.webshop.productservice.dto.attribute.AttributeUpdateDto;
import com.intexsoft.webshop.productservice.model.Attribute;
import org.mapstruct.InjectionStrategy;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingConstants;

import java.util.List;
import java.util.Set;

@Mapper(
        componentModel = MappingConstants.ComponentModel.SPRING,
        injectionStrategy = InjectionStrategy.CONSTRUCTOR
)
public interface AttributeMapper {
    @Mapping(target = "id", ignore = true)
    @Mapping(target = "label", source = "label")
    @Mapping(target = "subcategory", ignore = true)
    @Mapping(target = "attributeValues", ignore = true)
    Attribute toAttribute(AttributeCreateDto attributeCreateDto);

    @Mapping(target = "id", source = "id")
    @Mapping(target = "label", source = "label")
    @Mapping(target = "subcategory", ignore = true)
    @Mapping(target = "attributeValues", ignore = true)
    Attribute toAttribute(AttributeUpdateDto attributeUpdateDto);

    @Mapping(target = "id", source = "id")
    @Mapping(target = "label", source = "label")
    AttributeDto toAttributeDto(Attribute attribute);

    List<AttributeDto> toAttributeDtoList(Set<Attribute> attributes);
}

File: product-service-kt/src/main/kotlin/com/intexsoft/webshop/productservicekt/mapper/AttributeMapper.kt
package com.intexsoft.webshop.productservicekt.mapper;

import com.intexsoft.webshop.productservicekt.dto.attribute.AttributeCreateDto
import com.intexsoft.webshop.productservicekt.dto.attribute.AttributeDto
import com.intexsoft.webshop.productservicekt.dto.attribute.AttributeUpdateDto
import com.intexsoft.webshop.productservicekt.model.Attribute
import org.mapstruct.InjectionStrategy
import org.mapstruct.Mapper
import org.mapstruct.Mapping
import org.mapstruct.MappingConstants

@Mapper(
    componentModel = MappingConstants.ComponentModel.SPRING,
    injectionStrategy = InjectionStrategy.CONSTRUCTOR
)
interface AttributeMapper {
    @Mapping(target = "id", ignore = true)
    @Mapping(target = "label", source = "label")
    @Mapping(target = "subcategory", ignore = true)
    @Mapping(target = "attributeValues", ignore = true)
    fun toAttribute(attributeCreateDto: AttributeCreateDto): Attribute

    @Mapping(target = "id", source = "id")
    @Mapping(target = "label", source = "label")
    @Mapping(target = "subcategory", ignore = true)
    @Mapping(target = "attributeValues", ignore = true)
    fun toAttribute(attributeUpdateDto: AttributeUpdateDto): Attribute

    @Mapping(target = "id", source = "id")
    @Mapping(target = "label", source = "label")
    fun toAttributeDto(attribute: Attribute): AttributeDto

    fun toAttributeDtoList(attributes: MutableSet<Attribute>): MutableList<AttributeDto>
}